### ISSUE TITLE
docker cp into container does bad things

### ISSUE THREAD

**Issue by j-griffith** [moby/moby#34665]
**Description**
When using docker cp to copy data into a plugin provided volume docker never restores the original mount in the destination container.

**Steps to reproduce the issue:**
1. docker volume-create -d cinder foo
2. docker run -it --volume-driver cinder -v foo:/data --name bar debian bash
3. docker cp testfile.dat bar:/data/

The container looses it's attachment (/data is no longer available).  I have found that if using a device that uses shared targets (the iSCSI target is never deleted) then /data will still be accessible but the steps below might illustrate why.

**Describe the results you received:**
I've noticed based on the logs that there's some things that go on here that seem kinda strange.  When the `docker cp` is issued the sequence of events in the logs looks like this:
a.  Mount request is issued to the plugin for volume foo
    Odd given that the volume is already mounted and we specifically specified to utilize a mounted 
    volume in the container (bar:/data)

The plugin responds that the volume is already mounted at /var/lib/docker-volumes/cinder/foo

b. An Unmount request is issued to the plugin for volume foo
    Not sure why, but immediately after the above call an Unmount is requested from volume-plugin-
    helper.  In the case of devices that use unique targets for each volume on the device (tgt driver 
    and LVM for example) this Unmounts the device and it also detaches the iSCSI connection.

c. A Mount request is issued to the plugin for volume foo
    Same mount point, same request as before, nothing new here.. so we go ahead and just do it 
    again

d. Another Unmount request is issued to the plugin for volume foo
    This is the last call in logs from the docker cp process

I can issue a restart on the container; the volume is reattached and the data from the `docker cp` command is there and valid.

**Describe the results you expected:**
I expected to be able to issue the `docker cp` command and not have to restart my container, also frankly I wouldn't have expected all the Unmount/Mount business taking place at all.  The volume is mounted, why not just cp to the mountpoint without the Unmount/Mount dance inbetween?

**Additional information you deem important (e.g. issue happens only occasionally):**
I'm not quite sure what value the `docker cp` command provides anyway?  It's great to wrap thing and make them more convenient, maybe it's just to update data while a container/process is running?  In which case this is kinda dangerous because you're ripping out a device from a mounted FS while it's running.  

If you just want to initialize data it's rather easy to just do something like:
`docker run -v myfile.data:/src --volume-driver cinder -v foo:/dest debian cp -R /src/* /dest/*` or something along those lines?  If it is in fact to be able to dynamically update volume contents while running we shouldn't unmount/disconnect at the risk of corrupting the FS.

**Output of `docker version`:**

```
Client:
 Version:      17.06.1-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   874a737
 Built:        Thu Aug 17 22:51:12 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.1-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   874a737
 Built:        Thu Aug 17 22:50:04 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 17.06.1-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 3
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local solidfire
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 6e23458c129b551d5c9871e5174f6b1b7f6d1170
runc version: 810190ceaa507aa2727d7ae6f4790c76ec150bd2
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-93-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.953GiB
Name: ndvp-2
ID: GWDC:SF4K:76JZ:EE3S:4OZ5:O6W2:AGPK:VFOF:6KJ6:ZZLE:MZPF:6YSM
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
VM running Ubuntu 16.04
Also I've checked this with the Cinder plugin using TGT and LVM as well as the NetApp NDVP plugin using SolidFire and both exhibit the same behavior.

**Comment by cpuguy83** [comment]
Mount/Unmount are sent because docker cannot make any assumptions about what is mounted and what is not. It may be mounted in a container, but that does not mean it is mounted on the host.

Each Mount/Unmount pair uses a unique ID so plugins can reference count these mounts. However even if a plugin has a bug, an `Unmount` should not affect a running container since it's mount namespace is private... but maybe there's more than just mounts involved here.

**Comment by j-griffith** [comment]
@cpuguy83 
> Mount/Unmount are sent because docker cannot make any assumptions about what is mounted 
> and what is not. It may be mounted in a container, but that does not mean it is mounted on the 
> host.

That's perfectly reasonable and good, but if the plugin responds that the mount is in fact there and responds with the standard mount response appropriately why the subsequent Unmount/Mount?  I guess I haven't really thought about how you would have a volume in a container that wasn't mounted on the host, not sure how that looks but fair enough.

> Each Mount/Unmount pair uses a unique ID so plugins can reference count these mounts.

That's what threw me here, there are 3 Mount/Unmount request pairs, the first 2 pairs are in fact using the same Req ID.  I was just confused as to why the first Mount was followed by an Unmount to begin with?

> However even if a plugin has a bug, an Unmount should not affect a running container since it's 
> mount namespace is private... but maybe there's more than just mounts involved here.

So the thing I was running into is that with the unique targets (and the fact that I'm not doing ref-counting and obvioulsy I need to) when an Unmount for the volume was requested I am going through and performing an Unmount of the device from `/var/lib/docker-volumes/cinder/foo` (which is the same in my case regardless of Req ID) but what's worse is since I assume single attach I don't ref count I take that Unmount request to mean "we're done goodbye" and I do an iscsiadm logout and remove the target as well.

Let me know if you have any other thoughts about the first Mount/Unmount and if you think that's expected/normal or not.  Otherwise I guess I will close this out and implement the reference counting in the plugin.

Thanks for the insight!

**Comment by cpuguy83** [comment]
The first mount/unmount likely comes from the create phase of the container where docker copies all data from the image path that the volume will be mounted into into the volume (assuming the volume is empty).
Then at start it asks to mount again.

Note that `docker run` is the equivalent of `docker create` + `docker start`.

**Comment by j-griffith** [comment]
@cpuguy83 Sure, so actually I wasn't even mentioning the initial Mount request from when I started the container.  This is just after the container is up and running and I verified that the volume was there and accessible.  From the time I issue the docker cp command I see the Mount request, which I respond with already mounted and provide the regular mount response, and then I immediately recieve the Unmount followed by another Mount both of which are using the same Req ID.  Let me fix up the ref counting and make sure I'm not doing something else stupid here.

**Comment by j-griffith** [comment]
@cpuguy83 Ok, I added some simple reference counting and that did the trick.  There is something not right in the sequence of mount and unmount calls and the request ID's don't seem to necessarily match up with each other but regardless both the plugins I was working with needed to do the counting here so thanks for helping me out with the recommendation.  I'll go ahead and just close this as no issue.

**Comment by untoldone** [comment]
@cpuguy83 Shouldn't this behavior and requirement to reference count be documented at https://docs.docker.com/engine/extend/plugins_volume/#volumedrivercapabilities ? Other Volume Plugins have the same issues as a result. The only time I've seen this issue manifested is when using `docker cp` ... which suggests this is an edge case created by cp which is probably not tested very well by any volume plugins ... especially since the documentation on this behavior seems to be missing?

**Comment by cpuguy83** [comment]
@untoldone Do you have a suggestion for a change? Right now under `Mount` there is:

> Docker requires the plugin to provide a volume, given a user specified volume name. Mount is called once per container start. If the same volume_name is requested more than once, the plugin may need to keep track of each new mount request and provision at the first mount request and deprovision at the last corresponding unmount request.

**Comment by untoldone** [comment]
@cpuguy83 Let me try to fix the volume driver I'm using and then suggest something. I'm not sure I understand that paragraph to be honest.

I'm also still not sure I understand why mounts / unmounts are issued at all for a cp unless I don't understand the full scope of what cp can do.

If I issue a cp, I assume it is for a container that is already running and I'd also assume that the behavior is non-destructive to the state of the docker container configuration. I'm not sure why mounts / unmounts would be involved here at all.

That said, I'll share a suggestion once I fix the GCE plugin in the related issue I mentioned.

**Comment by j-griffith** [comment]
FWIW reference counting didn't solve anything here.  The fact is that the cp command when executed against a mountpoint in a container shouldn't go through the cycle of unmount,mount,unmount,mount,unmount at all.  Reference counting won't help here if you're treating connections as single target connections per-request because.

The thing I was trying to point out with this issue in the first place was that the behavior is not as described in the documentation for Mount.  The sequence and number of unmount calls results in the volume being unmounted even if you're doing reference counting.

I just solved this on my side by not removing iSCSI connections on unmounts until volume deletion.  This created some challenges with envs like Swarm but for now it's the best I could come up with.

**Comment by untoldone** [comment]
@j-griffith hmm your path doesn't sound like it will work in my case. I'm working with Google Compute Platform volumes that are remotely connected to instances.

Before digging to try to fix the plugin we're using, is it possible to know if a umount is for a `cp` vs for a container stop (not destroy)? umount while the container is running currently force detaches the volume with the plugin I'm using, but from what you described this doesn't even sound fixable without changes to docker. Is this understanding correct?

Just trying to get more info before spending the day fixing the plugin.

**Comment by cpuguy83** [comment]
1. `docker cp` works on not running containers as well.
2. The `docker cp` experience is expected to be exactly as it is in the container
3. Docker does not assume that something is mounted in its namespace.

If you are finding an uneven number of mounts/unmounts, this is a problem.
Every Mount(ID) should be followed by an Unmount(ID).

`docker create` -> mount +unmount
`docker start` -> mount
`docker stop` -> unmount
`docker cp` -> mount+unmount

Multiple containers can use a single volume, so you can also get two containers requesting the mount.

**Comment by untoldone** [comment]
@cpuguy83 the issue is that `docker cp` is issuing mounts and unmounts on a running docker container where the mounted filesystem is required for the container to continue operation. My expectation would be that if the container is running, no mounts / unmounts are issued. Calling mount / unmount in the case of the driver I'm working with (https://github.com/mcuadros/gce-docker) causes the disk to be forcibly removed which has caused filesystem corruption in my case by just issuing a `docker cp`... I assume the documentation you've referenced is basically saying that plugins need to know when mount needs to actually mount and unmount actually needs to umount.. basically the plugin needs to be stateful rather than stateless.

That said, running `cp` never runs `mount` or `unmount` if on a linux system which makes this behavior really unintuitive and the documentation on this behavior hard to follow.

Again, I'll take a closer look and try to suggest something. Thanks for your replies so far.

**Comment by cpuguy83** [comment]
> That said, running cp never runs mount or umount if on a linux system

What do you mean? The `cp` tool?
This is not the same thing as `docker cp`. `docker cp` needs to make sure all the filesystems are in place so that that what's visible to the `docker cp` caller is exactly what's in the container.

**Comment by cpuguy83** [comment]
Also yes, it would be rather difficult to write a stateless plugin for anything but simple cases.

**Comment by untoldone** [comment]
I understand they are not the same, but I would think you'd want them to be similar given their shared name in terms of people understanding what will happen.

That said, I also still don't understand why mount/ unmount would be issued for running cp. Why would docker remount when it's alread been mounted. Can't docker store that state itself?

Anyways, I'll see if I can fix the plugin but my 2c are this current model feels like most third party implementations of a volume plugin will be broken in this same way since the docs really don't seem clear here in needing to be aware of this pattern.

**Comment by cpuguy83** [comment]
Why should the plugin care what command requested a mount? If we stick to the spec and not how Docker works, then this should work. If it doesn't work with a correctly implemented spec then we need to look if Docker is doing wrong or if the spec is insufficient.

Docker could improve internally to not request mounts for things it knows are mounted, but in the end this shouldn't affect how the plugin operates.

**Comment by untoldone** [comment]
I think it makes sense for the plugin to not care what command issued the mount / umount -- I asked about that from @j-griffith as he mentioned reference counting didn't work for him. Assuming reference counting does work, my point is about how intuitive the spec is to implement.

My opinion here is that if Unmount is called, the volume should be unmounted. If mount is called, the volume should be mounted -- not that it should be unmounted only if there are any other outstanding calls to mount that hasn't been unmounted. Given that this actually means "Add one more to the number of things currently using the volume" it could have been named something more like "IssueMountHandle" or something else that could indicate we should increase a counter from the plugin side.

Given we have a spec already ... to me it feels like this should either be documented much more explicitly or change when docker itself calls mount / unmount -- only calling mount / unmount when docker actually wants a volume mounted or unmounted based.

Of these two options, the latter would be a non-breaking change given the current spec and I feel it would be much more usable from a plugin developer perspective and would negate the need for the additional documentation.

**Comment by j-griffith** [comment]
@untoldone 
> @j-griffith hmm your path doesn't sound like it will work in my case. I'm working with Google 
> Compute Platform volumes that are remotely connected to instances.
Oh!  Indeed, my approach isn't going to help you there.

> Before digging to try to fix the plugin we're using, is it possible to know if a umount is for a cp vs 
> for a container stop (not destroy)? 
I tried to determine a way to do this and didn't see any way to make that determination either from the existing code in the docker cp command nor when inspecting the incoming request.

@cpuguy83 & @untoldone 
Just to reiterate my initial comments regarding the sequence, and to illustrate the reference counter, I think this should work in your case @untoldone .  It does seem odd that we go through the sequence of things in this case but regardless as @cpuguy83 pointed it out it should probably solve your problem.  Of course I'm assuming you get the same sequence of calls that I saw:

Container is running with a volume attached
counter=1

1. Mount request is issued to the plugin for volume foo
That's cool, makes sense since cp command allows to run against a non-running container and all, and it's a benign operation anyway, my plugin just responds with the connection info (which based on the spec I believe is correct). 

Using a ref counter though this means counter++
counter=2

The plugin responds that the volume is already mounted at /var/lib/docker-volumes/cinder/foo

2. An Unmount request is issued to the plugin for volume foo
I don't know why this is issued?  But it is so you'd get counter--
counter=1

3. A Mount request is issued to the plugin for volume foo
Ok.. counter++
counter=2

4. Another Unmount request is issued to the plugin for volume foo
and now decrement, counter--
counter=1

So, we never hit zero so should've never detached and we have a ref count of 1 remaining which is what we'd expect.  The reason this didn't work for me is I was trying to keep my driver stateless and in my case I had the ability to modify things to use the volume delete approach.

**Comment by untoldone** [comment]
I didn't have too much trouble fixing the GCE volume plugin -- @j-griffith / @cpuguy83 Thanks for the help!

@cpuguy83 I'd say I now feel even more strongly that docker should be handling the mount / unmounts so there aren't ever mounts issued when the volume is already mounted. I bet every working volume plugin has very similar code around handling this or doesn't have code around the scenario at all. This would likely only show up as a problem when running `docker cp` in which case it becomes really hard to debug.

At a minimum, I would suggest being much more explicit with the documentation as I didn't figure this issue out without your help in this thread. I might suggest changing:

> Docker requires the plugin to provide a volume, given a user specified volume name. Mount is called once per container start. If the same volume_name is requested more than once, the plugin may need to keep track of each new mount request and provision at the first mount request and deprovision at the last corresponding unmount request.

To something like

> Docker requires the plugin to provide a volume, given a user specified volume name. Mount may also be called more than once for the same volume_name without an Unmount in between. This may happen if `docker cp` is used to copy off of a container that has the volume mounted or if more than one container requests the volume by name. It is therefore important for plugins to track the outstanding mount requests by ID and only deprovision the volume once an unmount request has been made for each outstanding mount request.

Either way, I hope this thread can help others in the future with issues around weird `docker cp` behavior causing filesystem corruption.

**Comment by untoldone** [comment]
Just a last thought here -- my volume driver runs in a container itself. My fixes only work if the volume driver container isn't reset as it stores its statefull-ness in memory. I could be storing the state of my volume driver to disk itself, but then it gets even more complicated for what I feel is a pretty simple task. Maybe I'm missing other aspects here, but I bet most volume plugins face this same problem.

Docker itself is the only thing that certainly lives as long as all the containers it hosts... to me it feels the ideal place to store the state of mounted volumes rather than delegating that to the plugins.

**Comment by j-griffith** [comment]
> I'd say I now feel even more strongly that docker should be handling the mount / 
> unmounts so there aren't ever mounts issued when the volume is already mounted.

What might be another good option is add a field in the response that says 'already-mounted' or something along those lines, the Mount IMO is easy to ignore and just say "ok", I kinda like that part.  It's the seemingly needless unmounts that cause me grief.

> Just a last thought here -- my volume driver runs in a container itself. My fixes only work if the 
> volume driver container isn't reset as it stores its mount statefull-ness in memory.

Yeah, this is why I opted NOT to use the ref counting mechanism.  I fully expect my driver container (or even if it's not containerized) to be able to deal with failures like this.

>  I bet every working volume plugin has very similar code around handling this or doesn't have code > around the scenario at all. 

When I was looking at this something I discovered poking around Github and talking to some folks was many of them are using shared targets, so they never actually remove the target until the volume is deleted, or a large number of others are using NFS mounts.  Anyway; I'd be curious about folks using ref counters with single target connections.  My question would be have they tested this command out?

Anyway, @untoldone sorry to throw in all the editorial and such, hopefully it was somewhat constructive and helpful.

@cpuguy83 depending on your thoughts regarding the idea around maintaining mount state in Docker or if the extra info in the Mount response sounds interesting I'd be happy to work on a PR or even start with a proposal?

**Comment by untoldone** [comment]
Your suggestion definitely works well too. I actually think this issue is a more significant one than I first realized yesterday and probably deserves its own bug that ought to be open currently.

Given this issue, I'll likely move our volume management to an external system like Kubernetes as I'm not confident a stable volume driver implementation is possible today with the use of `docker cp`... it's really weird to tell our team "`docker cp` is safe .. but only some of the time, just restart the container you want to copy from first to make sure outstanding mounts are currently accounted for correctly by our volume driver"

**Comment by cpuguy83** [comment]
I have yet to see why `docker cp` is an issue. I understand that storing state sucks, but it's something that has to be done, even on Kubernetes (I have implemented a Flexvolume driver).

Even if Docker does more ref counting on it's end, this doesn't fix anything at all.

**Comment by cpuguy83** [comment]
Also please have a look at github.com/container-storage-interface/spec
This is what we expect to implement for the next storage plugin spec in Docker.

**Comment by untoldone** [comment]
`docker cp` is the only realistic scenario in which someone would cause an Unmount api call without intending it to actually unmounting (e.g. you need to store state just for this). If you plan your infrastructure to now allow multiple containers having access to the same disk (e.g. for a database), you'll never see this come up otherwise.

Storing state itself is not hard ... that's not the problem (that said, it is still unintuitive still as to why it is something delegated to volume plugins). The problem is that we don't have all the information all of the time.

Here are two examples that detail the issue:
1. Volume plugin stores state in memory. The issue arises when a volume plugin is restarted but volumes Mount's with have not received all outstanding Unmounts yet. The volume plugin cannot currently recover knowledge of outstanding Mount IDs that haven't had Unmounts... the plugin does not know if Unmount actually should lead to a disk being disconnected / physically unmounted or not.
2. You decide to solve this by storing state to disk. When your volume plugin is restarted, you can rehydrate the state of all outstanding Mount IDs. This way when an unmount is issued, you know when you should actually de-provision a disk / physically unmount. The issue arises when Docker itself is restarted. Presumably, all outstanding Mount IDs are now invalid. Your Volume plugin doesn't know to invalidate these mounts and when a Mount followed by Unmount pair, which would have otherwise indicated a disk being physically unmounted -- the volume plugin still thinks there are outstanding Mount requests.

There are at least 2 solutions to 1 + 2 here. Both of these solutions we've proposed removes the need for volume plugins to track the state of Docker's "volume state machine". I think both of these solutions will dramatically simplify what is needed for a Docker volume plugin implementation.

**Comment by cpuguy83** [comment]
So I see your point, we can improve the plugin failure scenario by having docker track mounts. This seems like a reasonable thing to do since it does improve certain failure cases. It could probably be implemented [here](https://github.com/moby/moby/blob/master/volume/drivers/adapter.go#L161), but there are some weird cases where we need to be careful, particularly with live-restore.
 
I worry will hide actual real problems with the plugin implementation, though.
